lib/c-api/src/wasm_c_api/macros.rs

```diff
@@ -100,15 +103,19 @@ macro_rules! wasm_declare_vec {
                 ::std::mem::forget(bytes);
             }
 
-
             #[no_mangle]
-            pub unsafe extern "C" fn [<wasm_ $name _vec_delete>](ptr: *mut [<wasm_ $name _vec_t>]) {
-                let vec = &mut *ptr;
-                if !vec.data.is_null() {
-                    Vec::from_raw_parts(vec.data, vec.size, vec.size);
-                    vec.data = ::std::ptr::null_mut();
-                    vec.size = 0;
+            pub unsafe extern "C" fn [<wasm_ $name _vec_delete>](subject: Option<Box<[<wasm_ $name _vec_t>]>>) {
+                if subject.is_none() {
+                    return;
                 }
+
+                let subject: Box<[<wasm_ $name _vec_t>]> = subject.unwrap();
+
+                if !subject.data.is_null() {
+                    let _ = Vec::from_raw_parts(subject.data, subject.size, subject.size);
+                }
+
+                ::std::ptr::drop_in_place(Box::into_raw(subject));
```

Comment:
Without that, everything compiles nicely but we have a free error when executing a C program.